### PR TITLE
Conseil store balance updates

### PR DESCRIPTION
**Pull request by ivanopagano**
Add storage of balance updates along any operations that has any, at any level.

Add also the consumed gas property to any operation.

### CODE REVIEW

**Comment by piotrkosecki** [comment]
quick question - do you think this new table should be included in `TezosPlatformDiscoveryOperations` ?

**Comment by ba205** [comment]
Some follow-ups are necessary:

-We can actually obtain "balance_updates" from block_header_metadata as well as operations and operation result. We want to grab those, too, and create a new field in the balance_updates table called "source", which specifies whether or not this balance_update came from block_header_metadata, an operation, or an operation result.
-We want to update the blocks table with the following fields from block_header_metadata: "baker", "voting_period_kind", "nonce_hash", "consumed_gas", and all of the subfields of "level".
-We want to make tables called "proposals", "bakers", and "ballots". I will send you the schema.

**Comment by ivanopagano** [comment]
> Some follow-ups are necessary:
> 
> -We can actually obtain "balance_updates" from block_header_metadata as well as operations and operation result. We want to grab those, too, and create a new field in the balance_updates table called "source", which specifies whether or not this balance_update came from block_header_metadata, an operation, or an operation result.
> -We want to update the blocks table with the following fields from block_header_metadata: "baker", "voting_period_kind", "nonce_hash", "consumed_gas", and all of the subfields of "level".
> -We want to make tables called "proposals", "bakers", and "ballots". I will send you the schema.

Fine, I'll make the changes, shouldn't be too much fuss

**Comment by vishakh** [comment]
> We can actually obtain "balance_updates" from block_header_metadata as well as operations and operation result. We want to grab those, too, and create a new field in the balance_updates table called "source", which specifies whether or not this balance_update came from block_header_metadata, an operation, or an operation result.

Balance updates from the block header reflect rewards to bakers (miners) and penalties to bad actors in the system. Please keep in mind these balance updates are not linked an operation so the corresponding records will not have a foreign key into the `operations` table. We could introduce a foreign key into the `blocks` table to link block-based balance updates with blocks.

@ivanopagano @ba205

File: build.sbt [hunk 1]
```diff
@@ -5,6 +5,7 @@ val akkaHttpVersion = "10.1.0"
 val akkaVersion = "2.5.11"
 val slickVersion = "3.2.1"
 val circeVersion = "0.11.0"
+val catsVersion = "1.6.0-RC1"
```
Comment (piotrkosecki) [inline build.sbt]: why we want release candidate here? does it have any major features/fixes we need?

File: src/main/scala/tech/cryptonomic/conseil/tezos/DatabaseConversions.scala [hunk 2]
```diff
@@ -3,10 +3,15 @@ package tech.cryptonomic.conseil.tezos
...
 import tech.cryptonomic.conseil.tezos.FeeOperations._
 import tech.cryptonomic.conseil.util.Conversion
-import tech.cryptonomic.conseil.util.Conversion.Id
+import tech.cryptonomic.conseil.util.Conversion._
 
 object DatabaseConversions {
 
+  import java.sql.Timestamp
+
+  //BEWARE: implicit conversion!!
```
Comment (piotrkosecki) [inline src/main/scala/tech/cryptonomic/conseil/tezos/DatabaseConversions.scala]: please convince me why we need implicit conversion 😄

File: build.sbt (hunk 1 above)
Comment (ivanopagano) [reply build.sbt]: it simply was the latest available at the time, as suggested on the website.
Good for you to notice, as they've bumped to 1.6.0 stable, so I can update

File: src/main/scala/tech/cryptonomic/conseil/tezos/DatabaseConversions.scala (hunk 2 above)
Comment (ivanopagano) [reply src/main/scala/tech/cryptonomic/conseil/tezos/DatabaseConversions.scala]: low risk, high gain

File: src/main/scala/tech/cryptonomic/conseil/tezos/DatabaseConversions.scala [hunk 3]
```diff
@@ -3,10 +3,15 @@ package tech.cryptonomic.conseil.tezos
 import tech.cryptonomic.conseil.tezos.TezosTypes._
 import tech.cryptonomic.conseil.tezos.FeeOperations._
 import tech.cryptonomic.conseil.util.Conversion
-import tech.cryptonomic.conseil.util.Conversion.Id
+import tech.cryptonomic.conseil.util.Conversion._
 
 object DatabaseConversions {
 
+  import java.sql.Timestamp
```
[the remark this answers is not in the document]
Comment (ivanopagano) [reply src/main/scala/tech/cryptonomic/conseil/tezos/DatabaseConversions.scala]: in this case for no apparent reason, so I'll move it

File: src/main/scala/tech/cryptonomic/conseil/tezos/DatabaseConversions.scala [hunk 4]
```diff
@@ -15,6 +20,11 @@ object DatabaseConversions {
     case _ => None
   }
 
+  def extractBigDecimal(number: BigNumber): Option[BigDecimal] = number match {
+    case Decimal(value) => Some(value)
```
[the remark this answers is not in the document]
Comment (ivanopagano) [reply src/main/scala/tech/cryptonomic/conseil/tezos/DatabaseConversions.scala]: `Decimal`, it's a custom type containing the actual `BigDecimal` standard value.

The reason for having `TezosTypes.(Positive)BigNumber` and their corresponding subtypes is to
1. directly map the types defined on the tezos json schema
2. check validity of the values before decoding, hence the need to have a proper subtype that represents a json string that doesn't correctly represent a valid (positive) number, and is stored as a String

File: src/main/scala/tech/cryptonomic/conseil/tezos/DatabaseConversions.scala [hunk 5]
```diff
@@ -280,20 +360,26 @@ object DatabaseConversions {
...
-  def convertBlockAccountsAssociation(blockHash: BlockHash, blockLevel: Int, ids: List[AccountId]): List[Tables.AccountsCheckpointRow] =
-    ids.map(
-      accountId =>
-        Tables.AccountsCheckpointRow(
-          accountId = accountId.id,
-          blockId = blockHash.value,
-          blockLevel = blockLevel
-        )
-    )
+  implicit val blockAccountsAssociationToCheckpointRow = new Conversion[List, (BlockHash, Int, List[AccountId]), Tables.AccountsCheckpointRow] {
```
[the remark this answers is not in the document]
Comment (ivanopagano) [reply src/main/scala/tech/cryptonomic/conseil/tezos/DatabaseConversions.scala]: Should I? I didn't document any conversion as long as the meaning was straight.

I can amend that, but then it makes sense to document _all_ the conversions in this file, with an "arguably redundant" comment.

Should I proceed?

File: src/test/scala/tech/cryptonomic/conseil/tezos/JsonDecodersTest.scala [hunk 6]
```diff
@@ -37,6 +38,19 @@ class JsonDecodersTest extends WordSpec with Matchers with EitherValues {
 
     }
 
+    "decode a timestamp from ISO-8601 string format" in {
```
[the remark this answers is not in the document]
Comment (ivanopagano) [reply src/test/scala/tech/cryptonomic/conseil/tezos/JsonDecodersTest.scala]: Indeed, I wanted to be sure that the ISO format used by tezos for timestamps was correctly handled with Circe

🙂

File: src/main/scala/tech/cryptonomic/conseil/tezos/DatabaseConversions.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (vishakh) [reply src/main/scala/tech/cryptonomic/conseil/tezos/DatabaseConversions.scala]: Nope, your explanation makes sense.

**Pull request merged**